some refactoring to address Jeremy's concerns

pmario opened the pull request:
@cdruan The download button in the docs is always green now. .. It was too confusing, why it did change the color. .. I hope that is OK ... functionality wise

--------

The default syncer dropdown looks like this:

![extjs-syncer-01](https://user-images.githubusercontent.com/374655/121315641-d2e3d500-c908-11eb-9c89-4e9986cb62d8.gif)

Yours did look like this:

![extjs-syncer-02](https://user-images.githubusercontent.com/374655/121315213-68cb3000-c908-11eb-9f99-4e611b458bec.gif)

Mine now looks like this. It doesn't link to the help page anymore. It just downloads the core file, as expected.

![extjs-syncer-04](https://user-images.githubusercontent.com/374655/121315256-754f8880-c908-11eb-8a87-b43634d94818.gif)

 - The Default tiddler also opens a link to the "Using the external JavaScript template.tid "
 - The "Using the external JavaScript template.tid" was completely rewritten and merges the info from your snippet. 
   - I think it is more consistent now and contains all the info that is needed to reliably create a stand alone core.

pmario commented:
@cdruan Did you have time to review these changes? 

I think, your PR has great potential. Not only for WebDav. Also for configurations at the Azure free account, where "outbound" traffic seems to be limited and working with configurations that use a "copy on write" like filesystem like the former dat-project, that's now "hyper-xxx"

It would be really nice, if we could get it merged into TW 5.2.0 ... the next version

cdruan commented:
Hi @pmario,

Thanks for taking the time working on this! Yes I have reviewed your PR and I need a bit more time on the doc aspect of the PR.

With regard to the sync action, I'm afraid your version will not work because the downloaded file will be missing the crucial boot code as I tried to explain in the code comment. When core is externalized, the boot code that used to be contained in `$:/boot/boot.js`, etc. is now loaded by the <script> tag. So...

> Due to the essential boot code no longer exists as **tiddlers** when tiddlywiki starts up with the external-js template, the core JS needs to **come from the server** directly. 

When you use `tm-download-file` message, you are actually "exporting" core tiddlers within the wiki. If you try loading an external-js html with this tm-download-file version of twcore.js, you will find out that you get a blank page.

However, a link to the server's core JS with the "download" attribute will fire up a dialog box and save a working version core. The caveat is that older browsers like IE does not support the download attribute and will display a page full of JS instead... But I think the advantage probably outweighs the downside here. So I've modified the syncer dropdown using this technique.

I also took your comments and ideas. I moved the "export tiddlywiki core" button to the **Sidebar > Tools** section so `GetTiddlyWikiJS` tiddler can be removed without major impact. The problem with moving the export button to tw5.com edition is that the user may need a core that is older/newer than the tiddlywiki.com's version. In case it's confusing, this "export" button is different from the syncer dropdown in that it extract's the core from the wiki itself (i.e. your tm-download-file version). With this button, a standalone full html wiki can obtain its core without tiddlywiki on node.js.

I've pushed new commits that you may look over. All that's needed (I hope) is to edit the user doc on tw5.com.

pmario commented:
> Thanks for taking the time working on this! Yes I have reviewed your PR and I need a bit more time on the doc aspect of the PR.
> 
> With regard to the sync action, I'm afraid your version will not work because the downloaded file will be missing the crucial boot code as I tried to explain in the code comment. When core is externalized, the boot code that used to be contained in `$:/boot/boot.js`, etc. is now loaded by the <script> tag. So...

I trust you on this one. You did dig deeper into the mechanism as I did. 

> > Due to the essential boot code no longer exists as **tiddlers** when tiddlywiki starts up with the external-js template, the core JS needs to **come from the server** directly.
> 
> When you use `tm-download-file` message, you are actually "exporting" core tiddlers within the wiki. If you try loading an external-js html with this tm-download-file version of twcore.js, you will find out that you get a blank page.

OK. I see.

> However, a link to the server's core JS with the "download" attribute will fire up a dialog box and save a working version core. The caveat is that older browsers like IE does not support the download attribute and will display a page full of JS instead... But I think the advantage probably outweighs the downside here. So I've modified the syncer dropdown using this technique.

Even if IE shows a page full with JS code, you still can use the browser "save file as" dialogue. So the problem shouldn't be too big. 

> I also took your comments and ideas. I moved the "export tiddlywiki core" button to the **Sidebar > Tools** section so `GetTiddlyWikiJS` tiddler can be removed without major impact. The problem with moving the export button to tw5.com edition is that the user may need a core that is older/newer than the tiddlywiki.com's version. In case it's confusing, this "export" button is different from the syncer dropdown in that it extract's the core from the wiki itself (i.e. your tm-download-file version). With this button, a standalone full html wiki can obtain its core without tiddlywiki on node.js.

OK

> I've pushed new commits that you may look over. All that's needed (I hope) is to edit the user doc on tw5.com.

I think, the docs is much clearer now. ... BUT you have to have a look that the "big green button" again :)

pmario commented:
@cdruan I don't understand, what the button 
![grafik](https://user-images.githubusercontent.com/374655/121530318-9b0a8980-c9fd-11eb-94bf-e2fcacb09776.png) is 
used for in a single file TW. ... IMO it isn't necessary and confusing. I'm pretty sure Jeremy won't like it. 

For me the button if used from a single file wiki creates twcore-5.1.24-prerelease(2).js(1).html  .. Which is super strange

IMO We only need to be able to create the stand alone core in 3 cases

 - If we start from a file URI, which uses the ext-core
 - If we WebDav server client setting
 - If we start node-js version with the root tiddler set to ext-js

IMO There is no need for this button in a single file wiki.

pmario commented:
I think the my new documentation contains enough info for advanced users, to create what we want. ... If we "overengineer" it, it only creates confusion. ... I think, if you remove this button and use my docs, we would be good to go.

cdruan commented:
Hi @pmario 

> ... IMO it isn't necessary and confusing. I'm pretty sure Jeremy won't like it.

The "export" button and the "syncer" dropdown are set up to be mutually exclusive. When you can see "save TiddlyWiki core for offline use", the export button will be disabled (because it won't work). When the export button is enabled, the core is not externalized, and you won't see the syncer option.
 
> For me the button if used from a single file wiki creates twcore-5.1.24-prerelease(2).js(1).html .. Which is super strange

Do you think you can find out why? I can't reproduce it from my end. The underlying code is just tm-download-file message.

> IMO We only need to be able to create the stand alone core in 3 cases
> 
>     * If we start from a file URI, which uses the ext-core
> 
>     * If we WebDav server client setting
> 
>     * If we start node-js version with the root tiddler set to ext-js
> 
> 
> IMO There is no need for this button in a single file wiki.

This button is the easiest way for a FULL single file wiki to acquire the core--without having to install tiddlywiki on node.js. Once you have the core, you can convert any single file wiki into extxernal-js by change the value in `$:/config/SaveWikiButton/Template`. It will also be handy when you need to upgrade your external-js single file wiki in the future.

cdruan commented:
Hi @pmario 

OK, I've pushed updated doc and removed `$:/language/Snippets/GetTiddlyWikiJS` tiddler.

pmario commented:
>This button is the easiest way for a FULL single file wiki to acquire the core--without having to install tiddlywiki on node.js. Once you have the core, you can convert any single file wiki into extxernal-js by change the value in $:/config/SaveWikiButton/Template. It will also be handy when you need to upgrade your external-js single file wiki in the future.

Yea, but there is no "real intention" to advertise this possibility for ordinary users. There is a reason, why TWclassic, which basically is from 2004, is still used and maintained by some users. It's a single file. 

So this functionality will be for advanced users only and they should be able to create the ext-core if they need it. ...

The problem with the new "page-button" is, that if you click it, it will download the externalized core. ... 

Then newbie users will think: "What now". .. Since there is no answer they will start to ask, just to get the info: Don't use it! ... 

If the 2 file solution that both of us desperately want to have is done wrong, it will cause more problem than it solves. So the core should be simple. And if they need it they will start to ask: How can I create the 2 file solution and we can point them to the docs. .. IMO it should be enough for advanced users. 

The only thin you should do is, `tiddlywiki ./test --build listen` in tiddlywiki.info should also contain the `enable-cache=yes` flag, so it does make a difference.  See my PR's `tiddlywiki.info`

cdruan commented:
> The only thin you should do is, `tiddlywiki ./test --build listen` in tiddlywiki.info should also contain the `enable-cache=yes` flag, so it does make a difference. See my PR's `tiddlywiki.info`

Thanks, done.

> The problem with the new "page-button" is, that if you click it, it will download the externalized core. ...
>
> Then newbie users will think: "What now". .. Since there is no answer they will start to ask, just to get the info: Don't use it! ...

I made the hint more descriptive. It now says "Export tiddlywiki core for external-js wiki" and it's tugged away (down at the bottom of the list).

pmario commented:
>I made the hint more descriptive. It now says "Export tiddlywiki core for external-js wiki" and it's tugged away (down at the bottom of the list).

:) ... OK we will see, what Jeremy says

pmario commented:
I'd really like to have this PR merged, so I can try it with Azure

pmario commented on editions/server-external-js/tiddlywiki.info:
```diff
@@ -10,17 +10,21 @@
+			],
+		"listen": [
+			"--listen","root-tiddler=$:/core/save/all-external-js"
+			]
```
@cdruan ... This options should be moved to the end of the list, as shown here, in your upstream PR. 

If you do `tiddlywiki .xxxx --build`  it will start at the top and build all elements. .. If this option is the first one it will stop there. The rest will probably be built after the "listen" process stopped. .. I think this is a bit "wonky"

cdruan replied:
Wo! I didn't know about the missing target behavior of the --build command. Thanks, @pmario!

I will move "listen" target to the end of the list.

cdruan replied:
wait... I got curious and just tested what would happen if "listen" was listed first. It looks like tw still builds all of the targets before starting to listen. So I guess no change is needed.

pmario commented:
I'll close the pr. thx

cdruan commented:
Hi @pmario 

I just experimented with pulling out my PR into a plugin and I believe it worked! Would you be interested in that? I wonder if plugin is a better place for this, given that pulling the core out of a single-file wiki seems to be discouraged.

pmario commented:
I think it should be merge-able now and it is a valuable addition with a little bit of care taken by the user. I think it can be used in various ways. Whereas the initial implementation, that is already part of the core, is limited to exactly 1 usecase, nobody really needs. 

If it doesn't get merged I and others will be interested in the plugin.

pmario commented:
@cdruan ... There is 1 thing missing: in the [build-site.sh](https://github.com/Jermolene/TiddlyWiki5/blob/master/bin/build-site.sh) script we need this. ... 

May be after 
```
# /editions/tw5.com-docs/index.html	tiddlywiki.com docs edition
```

This will be needed

```
# /editions/server-external-js/index.html				Custom edition with externalized core
node $TW5_BUILD_TIDDLYWIKI \
	./editions/server-external-js\
	--verbose \
	--load $TW5_BUILD_OUTPUT/build.tid \
	--output $TW5_BUILD_OUTPUT \
	--build index \
```

This will allow us to do: 

```
tiddlywiki ./test --init server-external-js
tiddlywiki ./test --build listen
```

and skip the initial `--build index` step that was mentioned by Saq at: https://github.com/Jermolene/TiddlyWiki5/pull/5570#issuecomment-860619538

~~We may also consider, to rename the output file in tiddlywiki.info from `index.html` to `external-js.html` ... may be?!~~ .. IMO index.html is better.

cdruan commented:
Hi @pmario,

I am not seeing the connection between the two (build-site.sh and init new wiki)...

Furthermore, it looks like `build-site.sh` is for building the production site. I don't think I'm in the position to modify it.
